---
title: "SeqLogoPlot"
author: "Defeng Bai Agricultural Genomics Institute at Shenzhen, Chinese Academy of Agricultural Sciences"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo=T, comment=NA, message=F, warning=F,
	fig.align="center", fig.width=5, fig.height=3, dpi=300)
```

## 序列标识图
Sequence logo

序列标识图(sequence logo), 序列标识图就是序列的logo，它是以图形的方式依次绘出序列比对中各个位置上出现的残基，每个位置上残基的累积可以反映出该位置上残基的一致性。每个残基对应图形字符的大小与残基在该位置上出现的频率成正比。在生物信息分析中，经常要做序列标识图。

Sequence logo, a sequence logo is the logo of a sequence. It is a graphical representation of the residues at each position in the sequence alignment. The accumulation of residues at each position can reflect the consistency of the residues at that position. The size of the graphic character corresponding to each residue is proportional to the frequency of occurrence of the residue at that position. In bioinformatics analysis, sequence logos are often used.


## 序列标识图R语言实战
Sequence logo using R software

参考：https://mp.weixin.qq.com/s/dh1SoNJq3rnw5fuaPp98lw

```{r sequence logo}
# Install and load required packages
if (!requireNamespace("ggseqlogo", quietly = TRUE)) {
  install.packages("ggseqlogo")
}
if (!requireNamespace("cowplot", quietly = TRUE)) {
  install.packages("cowplot")
}

library(ggplot2)
library(ggseqlogo)
library(gridExtra)
library(cowplot)

# Some sample data
data(ggseqlogo_sample)

# Example data from ggseqlogo package
#data("pfms_dna")  # Matrix data
#data("seqs_dna")  # DNA sequences

# 1. Basic Logo Plot
p1 <- ggseqlogo(seqs_dna$MA0001.1, 
                method = "bits", 
                stack_width = 0.95) +
  ggtitle("Basic Logo Plot") +
  theme_logo(base_size = 14) +
  theme(legend.position = "bottom", 
        axis.text.x = element_text(size = 12),
        plot.title = element_text(size = 16))

# 2. Matrix Data Plot
p2 <- ggseqlogo(pfms_dna$MA0018.2, 
                method = "bits") +
  ggtitle("Matrix Data Plot") +
  theme_logo(base_size = 14) +
  theme(legend.position = "bottom", 
        axis.text.x = element_text(size = 12),
        plot.title = element_text(size = 16))

# 3. Bit vs Probability Method
p3 <- grid.arrange(
  ggseqlogo(seqs_dna$MA0001.1, method = 'bits') + ggtitle("Bits Method") +
    theme_logo(base_size = 14) +
    theme(legend.position = "bottom", 
          axis.text.x = element_text(size = 12),
          plot.title = element_text(size = 16)),
  
  ggseqlogo(seqs_dna$MA0001.1, method = 'prob') + ggtitle("Probability Method") +
    theme_logo(base_size = 14) +
    theme(legend.position = "bottom", 
          axis.text.x = element_text(size = 12),
          plot.title = element_text(size = 16)),
  
  ncol = 2
)

# 4. Custom Color Schemes
cs1 <- make_col_scheme(chars = c('A', 'T', 'C', 'G'), 
                       groups = c('group1', 'group1', 'group2', 'group2'), 
                       cols = c('purple', 'purple', 'blue', 'blue'))

cs2 <- make_col_scheme(chars = c('A', 'T', 'C', 'G'), 
                       values = 1:4)

p4 <- ggseqlogo(seqs_dna$MA0001.1, 
                col_scheme = cs1) +
  ggtitle("Custom Color Scheme") +
  theme_logo(base_size = 14) +
  theme(legend.position = "bottom", 
        axis.text.x = element_text(size = 12),
        plot.title = element_text(size = 16))

p5 <- ggseqlogo(seqs_dna$MA0001.1, 
                col_scheme = cs2) +
  ggtitle("Continuous Color Scheme") +
  theme_logo(base_size = 14) +
  theme(legend.position = "bottom", 
        axis.text.x = element_text(size = 12),
        plot.title = element_text(size = 16))

# 5. Logo with Annotations
p6 <- ggplot() +
  annotate('rect', xmin = 0.5, xmax = 3.5, ymin = -0.05, ymax = 1.9, 
           alpha = 0.1, col = 'black', fill = 'yellow') +
  geom_logo(seqs_dna$MA0001.1, stack_width = 0.90) +
  annotate('segment', x = 4, xend = 8, y = 1.2, yend = 1.2, size = 2) +
  annotate('text', x = 6, y = 1.3, label = 'Text Annotation') +
  ggtitle("Logo with Annotations") +
  theme_logo(base_size = 14) +
  theme(legend.position = "bottom", 
        axis.text.x = element_text(size = 12),
        plot.title = element_text(size = 16))

# 6. Font Variations
fonts <- list_fonts(F)
p7_list <- lapply(fonts, function(f) {
  ggseqlogo(seqs_dna$MA0001.1, font = f) + ggtitle(f) +
    theme_logo(base_size = 14) +
    theme(legend.position = "bottom", 
          axis.text.x = element_text(size = 12),
          plot.title = element_text(size = 16))
})

# Save plots to PDF
pdf("optimized_ggseqlogo_plots.pdf", width = 10, height = 12)

# Arrange plots in a grid
grid.arrange(p1, p2, p3, p4, p5, p6, ncol = 2)

# Print font variations
for (p in p7_list) {
  print(p)
}

# Close the PDF device
dev.off()

```


