---
title: "Circle packing plot"
author: "Defeng Bai Agricultural Genomics Institute at Shenzhen, Chinese Academy of Agricultural Sciences"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    toc_depth: 5
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo=T, comment=NA, message=F, warning=F,
	fig.align="center", fig.width=5, fig.height=3, dpi=300
)
```

### 圆堆积图
circle packing plot

什么是圆堆积图？
What is circle packing plot?

圆堆积图，circle packing plot是树形图的变体，使用圆形一层又一层的代表整个层次结构：树木的每个分支由一个圆圈表示，其子分支以圆圈内的圆圈来表示。每个圆圈的面积也可用来表示数量或文件大小等。也可用颜色将数据进行分类。

A circle packing plot is a variant of a tree diagram that uses circles to represent the entire hierarchical structure layer by layer: each branch of a tree is represented by a circle, and its sub branches are represented by circles within the circle. The area of each circle can also be used to indicate quantity or file size, etc. Color can also be used to classify data.


### 圆堆积图案例：
circle packing plot case

这是Jakob Stokholm课题组2023年发表于Nature Medicine上的文章，第一作者为Cristina Leal Rodríguez，题目为：The infant gut virome is associated with preschool asthma risk independently of bacteria

This is an article published by the Jakob Stokholm research group in Nature Medicine in 2023. The first author is Cristina Leal Rodr í guez, and the title is: The incident gut virus is associated with pre-school astroma risk independently of bacteria

![](e1.nature.medicine.2023.png)
Fig. 1 a, Circle packing visualization of the three distinct phage classes in the 1-year-old infant gut viromes. vOTUs were clustered at the protein level to define a taxonomic hierarchy. The outer circles represent the three main (caudoviral, microviral and inoviral) phage classes, and the inner circles, from darker to lighter colors, represent virus order-level clades (VOCs) and family-level clades (VFCs). Microviral and inoviral classes comprise only a single order each (Petitvirales and Tubulavirales, respectively). The size of each circle corresponds to the number of vOTUs embedded in each taxonomic group (n, total number of vOTUs per class).

图1 a，1岁婴儿肠道病毒中三种不同噬菌体类别的圆形包装可视化。vOTU在蛋白质水平上聚类以定义分类层次。外圈代表三种主要的噬菌体类别（尾病毒、微病毒和无病毒），内圈从较深到较浅的颜色代表病毒级分支（VOC）和家族级分支（VFCs）。微病毒类和非病毒类各仅包含一个单目（分别为小病毒类和管病毒类）。每个圆圈的大小对应于每个分类组中嵌入的vOTU的数量（n，每个类别的vOTu总数）。


### 圆堆积图R语言实现
Circle packing plot using R software
参考：https://mp.weixin.qq.com/s/5RizmZk7N9UyLGeI6so5Kw

```{r Circle packing plot, include=TRUE}
# 清除环境变量
# Clear environment variables
# rm(list = ls())

# 加载包
# Load packages
library(ggraph)
library(igraph)
library(tidyverse)
library(viridis)

# 构造数据
# Construct data
edges <- flare$edges
vertices <- flare$vertices
mygraph <- graph_from_data_frame(edges, vertices = vertices)

# 控制每个圆的大小
# Control the size of each circle
p1 <- ggraph(mygraph, layout = 'circlepack', weight = size) + 
  geom_node_circle() +
  theme_void()
p1
ggsave(p1, filename = '1.png', width = 5, height = 5)

# 根据深度上色
# Color according to depth
p2 <- ggraph(mygraph, layout = 'circlepack', weight = size) + 
  geom_node_circle(aes(fill = depth)) +
  theme_void() + 
  theme(legend.position = "none")
p2
ggsave(p2, filename = '2.png', width = 5, height = 5)

# 调色1
# Color palette 1
p3 <- p2 + scale_fill_viridis()
ggsave(p3, filename = '3.png', width = 5, height = 5)

# 调色2
# Color palette 2
p4 <- p2 + scale_fill_distiller(palette = "RdPu")
p4
ggsave(p4, filename = '4.png', width = 5, height = 5)

# 构造数据集的子集
# Constructing subsets of a dataset
edges <- flare$edges %>% 
  filter(to %in% from) %>% 
  droplevels()
vertices <- flare$vertices %>% 
  filter(name %in% c(edges$from, edges$to)) %>% 
  droplevels()
vertices$size <- runif(nrow(vertices))

# 构建graph对象
# Building a graph object
mygraph <- graph_from_data_frame(edges, vertices = vertices)

# 展示1
# Display 1
p5 <- ggraph(mygraph, layout = 'circlepack', weight = size) + 
  geom_node_circle(aes(fill = depth)) +
  geom_node_text(aes(label = shortName, filter = leaf, fill = depth, size = size)) +
  theme_void() + 
  theme(legend.position = "none") + 
  scale_fill_viridis()
p5
ggsave(p5, filename = '5.png', width = 5, height = 5)

# 展示2
# Display 2
p6 <- ggraph(mygraph, layout = 'circlepack', weight = size) + 
  geom_node_circle(aes(fill = depth)) +
  geom_node_label(aes(label = shortName, filter = leaf, size = size)) +
  theme_void() + 
  theme(legend.position = "none") + 
  scale_fill_viridis()
p6
ggsave(p6, filename = '6.png', width = 5, height = 5)

```

